src/lib.rs

```diff
@@ -176,6 +181,96 @@ impl<'t> Machine<'t> {
     }
 }
 
+/// Use this trait to launch machines into providers.
+///
+/// Important: You must call `cleanup` to shut down the instances once you are done. Otherwise, you
+/// may incur unexpected charges from the cloud provider.
+///
+/// This trait is sealed. If you want to implement support for a provider, see [`providers::Launcher`].
+pub trait Tsunami: sealed::Sealed {
+    /// A type describing a single instance to launch.
+    type MachineDescriptor: providers::MachineSetup;
+
+    /// Start up all the hosts.
+    ///
+    /// The returned future will resolve when the instances are spawned into the provided launcher.
+    /// SSH connections to each instance are accesssible via
+    /// [`connect_all`](providers::Launcher::connect_all).
+    ///
+    /// # Arguments
+    /// - `descriptors` is an iterator of machine nickname to descriptor. Duplicate nicknames will
+    /// cause an error. To add many and auto-generate nicknames, see the helper function
+    /// [`crate::make_multiple`].
+    /// - `max_wait` limits how long we should wait for instances to be available before giving up.
+    /// Passing `None` implies no limit.
+    ///
+    /// # Example
+    /// ```rust,no_run
+    /// #[tokio::main]
+    /// async fn main() -> Result<(), failure::Error> {
+    ///     use tsunami::Tsunami;
+    ///     // make a launcher
+    ///     let mut aws: tsunami::providers::aws::Launcher<_> = Default::default();
+    ///     // spawn a host into the launcher
+    ///     aws.spawn(
+    ///         vec![(String::from("my_tsunami"), Default::default())],
+    ///         None,
+    ///         None,
+    ///     )
+    ///     .await?;
+    ///     // access the host via the launcher
+    ///     let vms = aws.connect_all().await?;
+    ///     // we're done! terminate the instance.
+    ///     aws.cleanup().await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    fn spawn<'l>(
+        &'l mut self,
```

Comment:
I just realized that with the current signature, it's not possible to spawn some machines, then connect to them, and then spawn some more machines. Even though machines stay up for as long as the launcher lives once spawned. Not sure how to really fix this though...

Reply:
We'd have to make `spawn` take `&self`, and then use an internal `Mutex` or something to ensure that the spawns happen sequentially. Seems not worth it. And this should be pretty uncommon anyway. Feels more "right" to have it take `&mut self`. So I'll resolve for now.